pkg/repo/validate.go

```diff
@@ -14,53 +14,92 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package kepval
+package repo
 
 import (
-	"fmt"
 	"os"
 	"path/filepath"
 	"strings"
 
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 
-	"k8s.io/enhancements/api"
+	"k8s.io/enhancements/pkg/kepval"
 )
 
-const (
-	// TODO: Make this configurable and set in a client instead
-	DefaultPRRDir  = "prod-readiness"
-	kepMetadata    = "kep.yaml"
-	kepsReadmePath = "enhancements/keps/README.md"
-)
-
-var files = []string{}
-
 // This is the actual validation check of all KEPs in this repo
-func ValidateRepository(kepDir string) ([]string, map[string][]error, error) {
-	warnings := []string{}
-	valErrMap := make(map[string][]error)
+func (r *Repo) Validate() (
+	warnings []string,
+	valErrMap map[string][]error,
+	err error,
+) {
+	if r.ProposalPath == "" {
+		return warnings, valErrMap, errors.New("proposal path cannot be empty")
+	}
+
+	kepDir := r.ProposalPath
+	files := []string{}
 
 	// Find all the KEPs
-	err := filepath.Walk(kepDir, walkFn)
+	err = filepath.Walk(
+		kepDir,
+		func(path string, info os.FileInfo, err error) error {
+			logrus.Debugf("processing filename %s", info.Name())
+
+			if err != nil {
+				return err
+			}
+
+			if info.IsDir() {
+				if info.Name() == PRRApprovalPathStub {
+					return filepath.SkipDir
+				}
+
+				return nil
+			}
+
+			// true if the file is a symlink
+			if info.Mode()&os.ModeSymlink != 0 {
+				// Assume symlink from old KEP location to new. The new location
+				// will be processed separately, so no need to process it here.
+				return nil
+			}
+
+			dir := filepath.Dir(path)
+
+			metadataFilename := ProposalMetadataFilename
+			metadataFilepath := filepath.Join(dir, metadataFilename)
+			if _, err := os.Stat(metadataFilepath); err == nil {
+				// There is KEP metadata file in this directory, only that one should be processed.
+				if info.Name() == metadataFilename {
+					files = append(files, metadataFilepath)
+				}
+
+				return nil
+			}
+
+			if ignore(r, kepDir, info.Name()) {
+				return nil
+			}
+
+			// TODO(#2220): Return an error as soon as all KEPs are migrated to directory-based
+			//   KEP format.
+			files = append(files, path)
+			return nil
```

Comment:
With all KEPs migrated we're ready to do that (it's already done at head):
```
return fmt.Errorf("incorrect KEP format: %s", path)
```

(line 149)

Reply:
@wojtekt -- would you mind if we took that as a follow-up, since it requires some test refactoring that's sort of out of scope for this PR?

Reply:
@wojtek-t -- would you mind if we took that as a follow-up, since it requires some test refactoring that's sort of out of scope for this PR?

Reply:
If you could open an issue for it and replace the TODO with that issue, I would be fine with it.

Reply:
@wojtek-t -- I went a layer deeper and removed the legacy KEP logic in https://github.com/kubernetes/enhancements/pull/2528/commits/b8bdaf96a860d89179c67a34e72e717d3201cf74!

If the `len(files)` check after the filepath walk will handle the error if there are no KEPs in the directory and subdirectories:

```go
	if len(files) == 0 {
		return warnings, valErrMap, errors.New("must find more than zero KEPs")
	}
```

Reply:
This doesn't serve the purpose.

Imagine that in in directory "sig-foo", (where there are already N different KEPs) I'm adding a new KEP but using the old template. This won't be validated at all now. But we should error.

Reply:
@justaugustus ^^